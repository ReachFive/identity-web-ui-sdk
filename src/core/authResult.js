import { parseJwtTokenPayload } from '../helpers/jwt'
import { logError } from '../helpers/logger'

export function enrichAuthResult(response) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value) {
    return value && (value.accessToken || value.idToken);
}

export function isErrorResponse(value) {
    return value && value.error;
}
