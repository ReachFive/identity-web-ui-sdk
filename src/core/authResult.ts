import { AuthResult, ErrorResponse } from '@reachfive/identity-core';

import { parseJwtTokenPayload } from '../helpers/jwt';
import { logError } from '../helpers/logger';

export function enrichAuthResult(response: AuthResult) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload,
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value: unknown): value is AuthResult {
    return (
        value !== null &&
        typeof value === 'object' &&
        ('accessToken' in value || 'idToken' in value)
    );
}

export function isErrorResponse(value: unknown): value is ErrorResponse {
    return value !== null && typeof value === 'object' && 'error' in value;
}
