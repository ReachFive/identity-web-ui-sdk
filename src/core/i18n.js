import isPlainObject from 'lodash-es/isPlainObject';

export default function resolveI18n(defaultMessages, messages = {}) {
    const mergedMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key, params) => {
        const template = mergedMessages[key] || key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param]), template)
            : template;
    };
}

function flattenObject(object, prefix = []) {
    return isPlainObject(object)
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
