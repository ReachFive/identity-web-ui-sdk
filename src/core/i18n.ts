export type I18nMessages = Record<string, string>

export type I18nNestedMessages = Record<string, string | I18nMessages>

export type I18nMessageParams = Record<string, unknown>

export type I18nResolver = (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => string

export function resolveI18n<T extends (Record<string, T> | string)>(defaultMessages: I18nMessages = {}, messages: I18nNestedMessages = {}): I18nResolver {
    const mergedMessages: I18nMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => {
        const template = mergedMessages[key] ?? fallback?.(params) ?? key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param] as string), template)
            : template;
    };
}

export default resolveI18n

function flattenObject(object: I18nNestedMessages | string, prefix: string[] = []): Record<string, string> {
    return typeof object === "object" && object !== null
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
