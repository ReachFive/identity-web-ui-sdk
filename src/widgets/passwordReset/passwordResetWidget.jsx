import React from 'react';

import pick from 'lodash-es/pick';
import omit from 'lodash-es/omit';

import { parseQueryString } from '../../helpers/queryString'

import { Heading, Info, Link } from '../../components/miscComponent';
import { createMultiViewWidget } from '../../components/widget/widget';
import { PasswordEditorForm } from '../passwordEditor/passwordEditorWidget'

class MainView extends React.Component {
    static defaultProps = {
        onSuccess: () => { },
        onError: () => { }
    };

    handleSubmit = data => {
        return this.props.apiClient.updatePassword({
            ...omit(data, ['passwordConfirmation']),
            ...this.props.authentication
        });
    };

    onSuccess = () => {
        this.props.onSuccess();
        this.props.goTo('success');
    };

    render() {
        const { i18n } = this.props;

        return <div>
            <Heading>{i18n('passwordReset.title')}</Heading>
            <Info>{i18n('passwordReset.intro')}</Info>
            <PasswordEditorForm
                handler={this.handleSubmit}
                canShowPassword={this.props.canShowPassword}
                showLabels={this.props.showLabels}
                onSuccess={this.onSuccess}
                onError={this.props.onError} />
        </div>;
    }
}

const SuccessView = ({ i18n, loginLink }) => (
    <div>
        <Heading>{i18n('passwordReset.successTitle')}</Heading>
        <Info>{i18n('passwordReset.successMessage')}</Info>
        {loginLink && (
            <Info>
                <Link href={loginLink}>{i18n('passwordReset.loginLink')}</Link>
            </Info>
        )}
    </div>
);

const resolveCode = () => {
    const qs = (window.location.search && window.location.search.length)
        ? window.location.search.substr(1)
        : '';

    return { authentication: pick(parseQueryString(qs), ['verificationCode', 'email', 'purpose']) };
};

export default createMultiViewWidget({
    initialView: 'main',
    views: {
        main: MainView,
        success: SuccessView
    },
    prepare: options => ({
        ...options,
        ...resolveCode()
    })
});
