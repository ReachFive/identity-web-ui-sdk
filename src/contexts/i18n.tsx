import React, { ComponentType, PropsWithChildren, useMemo } from 'react';

import { I18nMessages, I18nNestedMessages, I18nResolver, resolveI18n } from '../core/i18n';

export interface Props {
  defaultMessages?: I18nMessages
  messages?: I18nNestedMessages
}

export const I18nContext = React.createContext<I18nResolver | undefined>(undefined);

export function useI18n(): I18nResolver {
  const context = React.useContext(I18nContext);
  if (!context) {
    throw new Error('No I18nContext provided');
  }

  return context;
}

export interface I18nProps {
  i18n: I18nResolver
}

export type WithI18n<P> = P & I18nProps

export function withI18n<T extends I18nProps = I18nProps>(WrappedComponent: ComponentType<T>) {
  const displayName = WrappedComponent.displayName || WrappedComponent.name || "Component";

  const ComponentWithI18n = (props: Omit<T, keyof I18nProps>) => {
    const i18n = useI18n();
    return <WrappedComponent {...({ i18n })} {...(props as T)} />;
  };

  ComponentWithI18n.displayName = `withI18n(${displayName})`;

  return ComponentWithI18n;
}

export function I18nProvider({ children, defaultMessages, messages }: PropsWithChildren<Props>): JSX.Element | null {
  const resolver = useMemo(() => resolveI18n(defaultMessages, messages), [defaultMessages, messages])
  return <I18nContext.Provider value={resolver}>{children}</I18nContext.Provider>
}
