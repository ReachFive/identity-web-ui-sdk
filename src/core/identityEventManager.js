import { enrichAuthResult } from './authResult'
import EventManager from '../helpers/eventManager'

export function createEventManager() {
    const eventManager = new EventManager();

    return {
        on(eventName, listener) {
            eventManager.on(eventName, listener);
        },

        off(eventName, listener) {
            eventManager.off(eventName, listener)
        },

        fireEvent(eventName, data) {
            if (eventName === 'authenticated') {
                eventManager.fire(eventName, enrichAuthResult(data));
            } else {
                eventManager.fire(eventName, data);
            }
        }
    }
}
