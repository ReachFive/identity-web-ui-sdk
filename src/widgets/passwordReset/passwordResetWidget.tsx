import React from 'react';

import { parseQueryString } from '../../helpers/queryString'

import { Heading, Info, Link } from '../../components/miscComponent';
import { createMultiViewWidget } from '../../components/widget/widget';
import { PasswordEditorForm, PasswordEditorFormData } from '../passwordEditor/passwordEditorWidget'
import { useReachfive } from '../../contexts/reachfive';
import { useRouting } from '../../contexts/routing';
import { useI18n } from '../../contexts/i18n';

interface MainViewProps {
    /**
     * Whether or not to provide the display password in clear text option.
     * @default false
     */
    canShowPassword?: boolean
    /**
     * Callback function called when the request has failed.
     */
    onSuccess?: () => void
    /**
     * Callback function called after the widget has been successfully loaded and rendered inside the container.
     * The callback is called with the widget instance.
     */
    onError?: () => void
    /**
     * Whether the form fields' labels are displayed on the form view.
     * @default false
     */
    showLabels?: boolean
}

const MainView = ({
    authentication,
    canShowPassword = false,
    onSuccess = () => {},
    onError = () => {},
    showLabels = false,
}: PropsWithAuthentication<MainViewProps>) => {
    const coreClient = useReachfive()
    const i18n = useI18n()
    const { goTo } = useRouting()

    const handleSubmit = ({ password }: PasswordEditorFormData) => {
        return coreClient.updatePassword({
            password,
            ...authentication
        });
    };

    const handleSuccess = () => {
        onSuccess();
        goTo('success');
    };

    return (
        <div>
            <Heading>{i18n('passwordReset.title')}</Heading>
            <Info>{i18n('passwordReset.intro')}</Info>
            <PasswordEditorForm
                handler={handleSubmit}
                canShowPassword={canShowPassword}
                showLabels={showLabels}
                onSuccess={handleSuccess}
                onError={onError} />
        </div>
    )
}

interface SuccessViewProps {
    loginLink?: string
}

const SuccessView = ({ loginLink }: SuccessViewProps) => {
    const i18n = useI18n()
    return  (
        <div>
            <Heading>{i18n('passwordReset.successTitle')}</Heading>
            <Info>{i18n('passwordReset.successMessage')}</Info>
            {loginLink && (
                <Info>
                    <Link href={loginLink}>{i18n('passwordReset.loginLink')}</Link>
                </Info>
            )}
        </div>
    )
}

const resolveCode = () => {
    const qs = window.location.search.substring(1)
    const { verificationCode, email } = parseQueryString(qs)
    return { authentication: { verificationCode, email } as Authentication };
};

type Authentication = { verificationCode: string,  email: string }
type PropsWithAuthentication<P> = P & { authentication?: Authentication }

export interface PasswordResetWidgetProps extends MainViewProps, SuccessViewProps {}

export default createMultiViewWidget<PasswordResetWidgetProps, PropsWithAuthentication<PasswordResetWidgetProps>>({
    initialView: 'main',
    views: {
        main: MainView,
        success: SuccessView
    },
    prepare: options => ({
        ...options,
        ...resolveCode()
    })
});
