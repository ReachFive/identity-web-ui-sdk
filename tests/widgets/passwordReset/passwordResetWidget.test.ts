/**
 * @jest-environment jsdom
 * @jest-environment-options {"url": "http://localhost/?email=alice@reach5.co&verificationCode=123456"}
 */

import { beforeEach, describe, expect, jest, test } from '@jest/globals';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom/jest-globals'
import 'jest-styled-components';

import { type Client } from '@reachfive/identity-core';

import { type I18nMessages } from '../../../src/core/i18n';
import type { Config } from '../../../src/types';

import passwordResetWidget from '../../../src/widgets/passwordReset/passwordResetWidget';

const defaultConfig: Config = {
    clientId: 'local',
    domain: 'local.reach5.net',
    sso: false,
    sms: false,
    webAuthn: false,
    language: 'fr',
    pkceEnforced: false,
    isPublic: true,
    socialProviders: ['facebook', 'google'],
    customFields: [],
    resourceBaseUrl: 'http://localhost',
    mfaSmsEnabled: false,
    mfaEmailEnabled: false,
    rbaEnabled: false,
    consentsVersions: {},
    passwordPolicy: {
        minLength: 8,
        minStrength: 2,
        allowUpdateWithAccessTokenOnly: true,
    }
};

const defaultI18n: I18nMessages = {}

describe('Snapshot', () => {
    const generateSnapshot = (options: Parameters<typeof passwordResetWidget>[0] = {}, config: Partial<Config> = {}) => async () => {
        // @ts-expect-error partial Client
        const apiClient: Client = {}

        const widget = await passwordResetWidget(options, {config: { ...defaultConfig, ...config }, apiClient, defaultI18n })
                
        await waitFor(async () => {
            const { container } = await render(widget);
            expect(container).toMatchSnapshot();
        })
    };

    describe('password-reset', () => {
        test('default', generateSnapshot());
    });
});

describe('DOM testing', () => {
    const updatePassword = jest.fn<Client['updatePassword']>()

    const onError = jest.fn()
    const onSuccess = jest.fn()
    
    beforeEach(() => {
        updatePassword.mockClear()
        onError.mockClear()
        onSuccess.mockClear()
    })

    const generateComponent = async (options: Parameters<typeof passwordResetWidget>[0] = {}, config: Partial<Config> = {}) => {
        // @ts-expect-error partial Client
        const apiClient: Client = {
            updatePassword
        }

        const result = await passwordResetWidget({ onError, onSuccess, ...options }, {config: { ...defaultConfig, ...config }, apiClient, defaultI18n });

        return render(result);
    };

    describe('password-reset', () => {
        test('basic', async () => {
            expect.assertions(6);
            
            const user = userEvent.setup()

            updatePassword.mockResolvedValue()

            await generateComponent();

            const password = screen.getByTestId('password')
            expect(password).toBeInTheDocument();
            
            const passwordConfirmation = screen.getByTestId('password_confirmation')
            expect(passwordConfirmation).toBeInTheDocument();
            
            const submitBtn = screen.getByTestId('submit')
            expect(submitBtn).toHaveTextContent('send');

            await user.type(password, 'azerty')
            await user.type(passwordConfirmation, 'azerty')
            await user.click(submitBtn)

            expect(updatePassword).toBeCalledWith({
                password: 'azerty',
                email: 'alice@reach5.co', /** @see @jest-environment-options on top of file */
                verificationCode: '123456', /** @see @jest-environment-options on top of file */
            })

            expect(onSuccess).toBeCalled()
            expect(onError).not.toBeCalled()
        });

        test('api failure', async () => {
            const user = userEvent.setup()

            updatePassword.mockRejectedValue('Unexpected error')

            await generateComponent();

            const password = screen.getByTestId('password')
            const passwordConfirmation = screen.getByTestId('password_confirmation')
            const submitBtn = screen.getByTestId('submit')

            await user.type(password, 'azerty')
            await user.type(passwordConfirmation, 'azerty')
            await user.click(submitBtn)

            expect(updatePassword).toBeCalled()

            expect(onSuccess).not.toBeCalled()
            expect(onError).toBeCalledWith('Unexpected error')
        });
    });
});
