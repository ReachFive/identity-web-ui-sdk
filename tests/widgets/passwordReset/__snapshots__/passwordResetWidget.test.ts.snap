// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot password-reset default 1`] = `
.c2 {
  margin-bottom: 15px;
  text-align: center;
  color: #212529;
  font-weight: bold;
  font-size: 16.8px;
}

.c3 {
  text-align: center;
  margin-bottom: 10px;
}

.c4 {
  color: #495057;
}

.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c1 {
  --color-primary: #229955;
  --color-destructive: #dc4e41;
  --color-background: #ffffff;
  --color-text: #495057;
  --color-border: #ced4da;
  --spacing-padding-y: 9;
  --spacing-padding-x: 12;
  --spacing-block-inner-height: 38;
  --spacing: 10;
  --font-generic: 14;
  --border-width: 1;
  --radius: 3;
}

.c9 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #ffffff;
  background-color: #229955;
  border: 1px solid #229955;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
}

.c9:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(34,153,85,0.5);
}

.c9:hover,
.c9:active {
  background-color: #1b7842;
  border-color: #1b7842;
}

.c9[disabled] {
  opacity: .65;
}

.c5 {
  position: relative;
}

.c7 {
  color: #495057;
  margin-bottom: 5px;
  display: none;
}

.c6 {
  margin-bottom: 10px;
}

.c6 > label::after {
  content: "\\A0*";
  color: #dc4e41;
}

.c8 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.428571429;
  color: #495057;
  border-radius: 3px;
  border-width: 1px;
  border-style: solid;
  border-color: #ced4da;
  background-color: #fff;
  background-image: none;
  padding: 9px 12px;
  -webkit-transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c8:focus {
  border-color: #5fdb94;
  box-shadow: 0 0 0 3px rgba(95,219,148,0.5);
  outline: 0;
}

.c8:disabled,
.c8[readonly] {
  background-color: #e9ecef;
}

.c8::-webkit-input-placeholder {
  color: #868e96;
}

.c8::-moz-placeholder {
  color: #868e96;
}

.c8:-ms-input-placeholder {
  color: #868e96;
}

.c8::placeholder {
  color: #868e96;
}

<div>
  <div
    class="c0 c1 r5-widget"
    transition="entering"
  >
    <div>
      <div
        class="c2"
      >
        passwordReset.title
      </div>
      <div
        class="c3 c4"
      >
        passwordReset.intro
      </div>
      <form
        class="c5"
        novalidate=""
      >
        <div
          class="c6"
          required=""
        >
          <label
            class="c7"
            for="password"
          >
            newPassword
          </label>
          <div
            style="position: relative;"
          >
            <input
              autocomplete="new-password"
              class="c8"
              data-testid="password"
              id="password"
              name="password"
              placeholder="newPassword"
              required=""
              title="newPassword"
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          class="c6"
          required=""
        >
          <label
            class="c7"
            for="password_confirmation"
          >
            passwordConfirmation
          </label>
          <div
            style="position: relative;"
          >
            <input
              autocomplete="new-password"
              class="c8"
              data-testid="password_confirmation"
              id="password_confirmation"
              name="password_confirmation"
              placeholder="passwordConfirmation"
              required=""
              title="passwordConfirmation"
              type="password"
              value=""
            />
          </div>
        </div>
        <button
          class="c9"
          data-testid="submit"
          type="submit"
        >
          send
        </button>
      </form>
    </div>
  </div>
</div>
`;
